File: Qt e Progetti/Progetti/2017-2018 Progetto Calcolatrice/Polygonator - Menon e Dalla Via/Java/src/Model/Polygons/Kite.java
package Model.Polygons;

import Model.Line;
import Model.Point;

import java.util.List;

public class Kite extends Quadrilateral {
    public Kite(final List<Point> vertices) throws Exception {
        super(vertices);
    }

    /**
     * Returns whether a list of points is a kite or not
     */
    public static boolean canBeCreatedWith(final Polygon polygon) throws Exception {
        if (!Quadrilateral.canBeCreatedWith(polygon) || polygon.isConcave()) return false;

        int adjacent = 0;

        for (int i = 0; i < polygon.getVertices().size(); i++) {
            Point p = i == 0 ? polygon.getVertices().get(polygon.getVertices().size() - 1) : polygon.getVertices().get(i - 1);
            Point n = i == polygon.getVertices().size() - 1 ? polygon.getVertices().get(0) : polygon.getVertices().get(i + 1);
            Point a = polygon.getVertices().get(i);

            if (new Line(p, a).equals(new Line(a, n))) adjacent++;
        }

        return adjacent >= 2;
    }

    @Override
    public double getArea() throws Exception {
        List<Line> d = getDiagonals();
        return d.get(0).getDistance() * d.get(1).getDistance() / 2;
    }

    @Override
    public double getPerimeter() throws Exception {
        double a = Line.getDistance(vertices.get(0), vertices.get(1));
        double b = Line.getDistance(vertices.get(3), vertices.get(0));
        if (Math.abs(a - b) < 0.0001) b = Line.getDistance(vertices.get(1), vertices.get(2));
        return (a + b) * 2;
    }

    @Override
    public String toString() {
        return "Kite";
    }
}

File: Qt e Progetti/Progetti/2017-2018 Progetto Calcolatrice/Polygonator - Menon e Dalla Via/C++/Model/Polygons/Kite.cpp
#include "Kite.h"

Kite::Kite(const list<Point> &vertices) : Quadrilateral(vertices) {}

bool Kite::canBeCreatedWith(const Polygon &polygon) {
    if (!Quadrilateral::canBeCreatedWith(polygon) || polygon.isConcave()) return false;

    int adjacent = 0;
    for (auto it = polygon.getVertices().begin(); it != polygon.getVertices().end(); it++) {
        Point p = it == polygon.getVertices().begin() ? *prev(polygon.getVertices().end()) : *prev(it);
        Point n = it == prev(polygon.getVertices().end()) ? *polygon.getVertices().begin() : *next(it);

        if (Line(p, *it) == Line(*it, n)) adjacent++;
    }
    return adjacent >= 2;
}

double Kite::getArea() const {
    list<Line> d = getDiagonals();
    return d.begin()->getDistance() * next(d.begin())->getDistance() / 2;
}

double Kite::getPerimeter() const {
    double a = Line::getDistance(*getVertices().begin(), *next(getVertices().begin()));
    double b = Line::getDistance(*getVertices().begin(), *prev(getVertices().end()));
    if (fabs(a - b) < 0.0001) b = Line::getDistance(*next(getVertices().begin()), *next(next(getVertices().begin())));
    return (a + b) * 2;
}

string Kite::getName() const {
    return "Kite";
}
